### PR TITLE
#819 Part2 : Scavenge back hotspots to add to consensus calls

### PR DESCRIPTION
**Pull request by kgaonkar6**
🚨 merge #1050 before this PR 

#### What scientific question is your analysis addressing?
Combine hotspots from each caller (strelka,mutect,vardict and lancet) and merge with consensus maf file.

#### What was your approach?
I gathered filtered RDS from `scratch/hotspot-detection` and filtered the combined hotspot calls to only calls uniquely present and not in consensus maf generated as part of `snv-caller` module.

This dataframe now contains unique calls that need to be scavenged back,  however the following columns are have read support and quality assignment which can be unique to each caller since they are calculated as part of the variant calling process. The values in these columns are substituted by mean values to create  1 unique row for that call. 
unique_cols <- c("t_depth",
                 "n_depth",
                 "t_ref_count",
                 "n_ref_count",
                 "t_alt_count",
                 "n_alt_count",
                 "caller",
                 "vcf_qual")

Now that we have these scavenged hotspots I merge them with consensus maf which is the final result required from this analysis.

#### What GitHub issue does your pull request address?
#819 

#### Which areas should receive a particularly close look?
- Currently I believe, taking the mean values of the read support and quality seems ok please let me know if you would handle it any other way.
- Some hotspots are supported by <5 t_alt_depths which might need a review?

#### Is there anything that you want to discuss further?
The final result of #819 would be a maf format file with scavenged hotspots merged with consensus maf file, but because it is larger than the limit it wont get pushed to remote. I believe eventually this file will be part of the download data, thought?

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes

#### What types of results are included (e.g., table, figure)?
table

#### What is your summary of the results?
Hotspots will be added to consensus maf calls in a next step where we gather hotspot calls missed in the consensus maf.

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [ ] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [ ] The analytical code is documented and contains comments.

### CODE REVIEW

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (jashapiro) [inline analyses/hotspots-detection/01-create-hotspot-maf.Rmd:179]: I think (unless this is a new feature) that write_tsv handles compression automatically.
```suggestion
  write_tsv(output_file, na = ".")
```

**Review by jashapiro — COMMENTED** [review]
Sorry for the delay in getting to this!

I have some initial comments, as well as some line level comments that came up as I looked through the code. I expect those are not complete, but just some things that I noticed along the way; I will look in more detail after we decide on some of the higher level discussions here. 

I realized looking at this that I had forgotten about some aspects of the callers that we will want to address before building a consensus. 

All strelka calls are listed as single nucleotides, while other calls can be multinucleotide calls. To deal with this, we split all MNVs to consecutive SNVs before joining tables: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/52f64f3debc7763520ac64ff685c69a7dd68bb2d/analyses/snv-callers/scripts/02-merge_callers.R#L138-L140

The relevant splitting code is here, and you should feel free to use that script if it works: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/analyses/snv-callers/util/split_mnv.R

But I just looked at that code, and noticed that it does not handle indels at all, which seems like a flaw! I will file a separate issue about this. Whatever we decide, we should implement the same methods in both places.

Some of this may affect the whole discussion about indels in #956 as well (https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/956#discussion_r602394001), though I think we can be probably follow the permissive definition I outlined there to generate the subsets we are reading in here, and then do the logic of joining them in this PR.

Some quick thoughts: we do not necessarily expect the start and end to match perfectly from each caller, so this is the main thing we will want to think about. Do we retain only the overlapping bases? If so, how do we handle that for insertions? (This is actually a more general problem with no good solution, unfortunately. What if one caller calls a `CAT -> CATAT` change as an insertion of `TA` and another calls it as an insertion of `AT`?)

Other thoughts: 

RE: using the mean for stats. In the SNV consensus caller, we used the depth & count statistics from `strelka`. Part of this was because we were requiring all callers to call a location, so we had the liberty to just choose a "standard", but we also wanted to be as consistent as possible, and taking averages has the potential to cause extra variation. I believe that in an earlier version (when we were testing 2 of 3 consensus) we had used `lancet` when `strelka` was not available. The relevant code is here (it is a bit opaque, but only the "join" columns are kept from the other tables, then strelka is added on, so all other values come from there): https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/52f64f3debc7763520ac64ff685c69a7dd68bb2d/analyses/snv-callers/scripts/02-merge_callers.R#L132-L136 

I would probably advocate for a similar approach here, for consistency. We might establish a a hierarchy of callers, and prefer the call from strelka, then mutect2, then lancet. 

I would strongly argue against including calls that are only made by VarDict, and I think based on the results I saw here that we can probably avoid all single-caller calls. (I do wonder how many single-caller calls are variations in indel locations though: it looked like many are indels).

Similarly, the consensus MAF file we generated was filtered by VAF. https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/52f64f3debc7763520ac64ff685c69a7dd68bb2d/analyses/snv-callers/scripts/02-merge_callers.R#L160-L167 
I would expect we should do the same thing here.

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: added to read_tsv for consensus file

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: this was perfect, thanks for the suggestion

**Comment by kgaonkar6** [comment]
Thanks @jashapiro  for the reviiew. I added in the split_mnv() function to split MNV in hotspots and some other suggestion from your review. 

I haven't remove vardict only calls yet do we want to filter only vardict calls them and has low read support? Asking mainly because only vardict called this mutation in BRAF in BS_F969DNHD which is low grad glioma  and thus BRAF is of interest here and also has enough read support.

Hugo_Symbol | Entrez_Gene_Id | NCBI_Build | Chromosome | Start_Position | End_Position | Strand | Variant_Classification | Variant_Type | Reference_Allele | Tumor_Seq_Allele1 | Tumor_Seq_Allele2 | Tumor_Sample_Barcode | t_alt_count
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
BRAF | 673 | GRCh38 | chr7 | 140753325 | 140753338 | + | In_Frame_Del | DEL | ATCGAGATTTCACT | ATCGAGATTTCACT | CATCA | BS_F969DNHD | 5
BRAF | 673 | GRCh38 | chr7 | 140753335 | 140753335 | + | Missense_Mutation | DNP | C | C | T | BS_NHRNZRQV | 22
BRAF | 673 | GRCh38 | chr7 | 140753336 | 140753336 | + | Missense_Mutation | DNP | A | A | T | BS_NHRNZRQV | 22

**Comment by cansavvy** [comment]
> I haven't remove vardict only calls yet do we want to filter only vardict calls them and has low read support? Asking mainly because only vardict called this mutation in BRAF in BS_F969DNHD which is low grad glioma  and thus BRAF is of interest here and also has enough read support.
> 

I would still find it highly suspect if none of the other callers thought it had enough support to call it. Summary of [VarDict conclusion described here](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/analyses/snv-callers/README.md)
Plus the rationale for this is difficult describe if in some cases we listen to VarDict and some we don't.

**Comment by jharenza** [comment]
Would there be any reason VarDict might be better at calling certain types of MNVs/INDELS than the other callers? @jashapiro @cansavvy @migbro @yuankunzhu

**Comment by jashapiro** [comment]
> Would there be any reason VarDict might be better at calling certain types of MNVs/INDELS than the other callers? 

I am sure there are! But we would have to weigh that against the fact that VarDict calls an order of magnitude more _everything_. In my mind, our goal should not (can not) be a perfect match to expected mutations; we will have to accept some level of false positives and false negatives across the data set. My bias is that we should look at the effect of the selection scheme globally, not necessarily based on trying to be sure we capture specific variants in specific samples.

That said, if we want to have a very permissive scheme, we can do that, but I will refer back to my initial comment on this analysis: https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/819#issuecomment-712881337 If the standards we are applying differ dramatically from those used for the initial consensus, we should not add these variants to the consensus calls, but rather to produce a _separate_ analysis of hotspot variants where we apply a different set of standards. (Note: In this case, the steps where the previous consensus calls are excluded would be removed  from this analysis.)

**Comment by jaclyn-taroni** [comment]
> In my mind, our goal should not (can not) be a perfect match to expected mutations; we will have to accept some level of false positives and false negatives across the data set. My bias is that we should look at the effect of the selection scheme globally, not necessarily based on trying to be sure we capture specific variants in specific samples.
>
> That said, if we want to have a very permissive scheme, we can do that, but I will refer back to my initial comment on this analysis: #819 (comment) If the standards we are applying differ dramatically from those used for the initial consensus, we should not add these variants to the consensus calls, but rather to produce a separate analysis of hotspot variants where we apply a different set of standards. 

I strongly agree with both of these points.

**Comment by migbro** [comment]
So, I'd say, in terms of what might be lacking by excluding Vardict is specifically MNVs. This is because strelka2 does not call them, and also if you recall, lancet is run only on exome regions + calls made by strelka2 and mutect2. So that means, at best, lancet can call MNVs in support of mutect2. However, since I read somewhere that the openPBTA process breaks down MNVs, perhaps this is less of a problem. The downside of course is that for those broken up MNVs, the coding consequence isn't quite accurate. Having said all that, MNVs are a very small number of calls, so I can see in your case the hesitancy to bring in Vardict, a FP beast, to shore this issue up.

**Comment by jashapiro** [comment]
> So, I'd say, in terms of what might be lacking by excluding Vardict is specifically MNVs. This is because strelka2 does not call them, and also if you recall, lancet is run only on exome regions + calls made by strelka2 and mutect2. So that means, at best, lancet can call MNVs in support of mutect2. However, since I read somewhere that the openPBTA process breaks down MNVs, perhaps this is less of a problem. The downside of course is that for those broken up MNVs, the coding consequence isn't quite accurate. Having said all that, MNVs are a very small number of calls, so I can see in your case the hesitancy to bring in Vardict, a FP beast, to shore this issue up.

I thought strelka2 called MNVs as consecutive SNVs? So in theory it should capture them at least by position, though as you say it loses information about the actual coding changes. You are correct that our consensus script splits MNVs from mutect & lancet before merging.

**Comment by migbro** [comment]
Yes, that is correct. It's sort of semantics really in that, they are consecutive SNV calls that any rational person would look at and say, wait a minute, isn't than a MNV?!

**Comment by jharenza** [comment]
For the sake of this project, it sounds like consensus will be to not use any VarDict only calls. For the ones we found that are most functionally relevant (_H3F3A_ and _BRAF_), we will do some digging on our end and try to determine if they were called in other callers in pre-pass (_H3F3A_ K28M for BS_EE73VE7V was found in the Strelka2 pre-pass VCF, but flagged as `LowEVS`), but not urgently.

**Comment by kgaonkar6** [comment]
Thanks @jashapiro @jharenza @migbro for the discussion. I updated the code to implement the following suggestions from the initial comment:
>I would probably advocate for a similar approach here, for consistency. We might establish a a hierarchy of callers, and prefer the call from strelka, then mutect2, then lancet.
>Similarly, the consensus MAF file we generated was filtered by VAF.
 
With the code here:
https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/83ebe2958fb3d2e6b18f91afdb44d10e86137c89/analyses/hotspots-detection/01-create-hotspot-maf.Rmd#L178-L188
I used coalesce() here to find the first non-missing value in a given a set of vectors to use the first NULL values within the order of strelka2 > mutect2> lancet > vardict

>I would strongly argue against including calls that are only made by VarDict, and I think based on the results I saw here that we can probably avoid all single-caller calls. (I do wonder how many single-caller calls are variations in indel locations though: it looked like many are indels).

I've updated the code to filter out vardict-only calls as the last step in the notebook here
https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/db0fdf7771daa58e28ce64cab7da75be08573956/analyses/hotspots-detection/01-create-hotspot-maf.Rmd#L224-L228

**Comment by kgaonkar6** [comment]
@jharenza  @jashapiro I've also removed the code snippet that was merging the hotspots with consensus maf as discussed in https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/819#issuecomment-712881337. 

Just wanted to summarize that the output `pbta-snv-scavenged-hotspots.maf.tsv.gz` in this module is now hotspots that are currently not in pbta-snv-consensus-mutation.maf.tsv.gz which are found in any caller but if it is unique to vardict those calls are not retained.

**Comment by jashapiro** [comment]
Thanks for the updates @kgaonkar6. I would like to have a bit of additional discussion about what the files here should look like, and I think we need to add clear documentation about them for users and downstream analysis. 

When I mentioned having a hotspot only file earlier on this PR, https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/961#issuecomment-819502619, I also suggested that we might want to have this file contain _all_ of the hotspot calls (including those that overlap conesnsus calls). My perspective there was that somebody might want to look at hotspots in isolation, and to do that they would ideally be able to get all the calls in a single file, produced with the same procedure.

The challenge is that if we do that there will be duplicated calls between the hotspot file and the consensus (we currently exclude consensus calls). This is of course surmountable, but requires documentation to tell people not to simply concatenate the two files. 

That said, this is also an opportunity to check that the calls (and annotation) are the same between the two methods when the consensus calls exist. I think that may be a check worth doing.


On the subject of filtering out vardict-only calls: how many of the 79 single-caller calls were due to vardict?

**Comment by jharenza** [comment]
> When I mentioned having a hotspot only file earlier on this PR, [#961 (comment)](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/961#issuecomment-819502619), I also suggested that we might want to have this file contain _all_ of the hotspot calls (including those that overlap conesnsus calls). 

Hi @jashapiro - I must have missed this as well. I feel like if we make the hotspot file a total hotspot MAF, then I think we will have to make a module for combining them prior to analyses, which gets at having a combined file in storage somewhere again - unless we are okay with having this redundant code in many downstream modules (eg - all subtyping modules, oncoprint, somatic interactions, etc). I would have ideally liked to have a consensus + hotspot MAF, but with the separation of files, I was imagining that we would have only the missed (scavenged) hotspots in the MAF, with the same columns used in the consensus MAF, such that the first step of any of the above mentioned analyses would be a simple pull header and cat the rest. I think this still gets at the "intentionality" of combining the files without making it overly difficult for people to merge these files. 

>My perspective there was that somebody might want to look at hotspots in isolation, and to do that they would ideally be able to get all the calls in a single file, produced with the same procedure.

Regarding this - the new MAFs in #1026 will have a new column called `Hotspot Allele` which denotes whether a call is a hotspot. Do you think that this would satisfy a user wanting to look at all hotspots in isolation? Presumably, if they would like to do such a thing, they may want them all, regardless of whether they were called by 2+/4 callers.

Thoughts?

**Comment by jashapiro** [comment]
> I feel like if we make the hotspot file a total hotspot MAF, then I think we will have to make a module for combining them prior to analyses, which gets at having a combined file in storage somewhere again - unless we are okay with having this redundant code in many downstream modules (eg - all subtyping modules, oncoprint, somatic interactions, etc). I would have ideally liked to have a consensus + hotspot MAF, but with the separation of files, I was imagining that we would have only the missed (scavenged) hotspots in the MAF, with the same columns used in the consensus MAF, such that the first step of any of the above mentioned analyses would be a simple pull header and cat the rest. I think this still gets at the "intentionality" of combining the files without making it overly difficult for people to merge these files.

No, I would not want to have modules combining the files prior to analysis all over the repo. My main point is that we should encourage and demonstrate using the different data for the purpose for which it was generated, and not just mix them together unless required. 

So for most (all?) subtyping modules, I would think that only the hotspot data would be needed. I am assuming that the genes that we are looking at in subtyping are all in the hotspot list, but if they are not, shouldn't they be?

For interactions, I think this will actually make the separation of analyses that are differentially affected by selection criteria (hotspot vs not) easier and more explicit. (This partly answers a separate question from earlier: how do we deal with bias from comparing sets with differential criteria? Keeping analyses explicitly separate is a good first step.)

> > My perspective there was that somebody might want to look at hotspots in isolation, and to do that they would ideally be able to get all the calls in a single file, produced with the same procedure.
> 
> Regarding this - the new MAFs in #1026 will have a new column called `Hotspot Allele` which denotes whether a call is a hotspot. Do you think that this would satisfy a user wanting to look at all hotspots in isolation? Presumably, if they would like to do such a thing, they may want them all, regardless of whether they were called by 2+/4 callers.

No, I don't think this is the same. This may be useful for some analyses (do hotspot variants look different in the consensus file, for example), but it means that if people want to look at just the hotspots, they have to read in a file that is much larger than they need,  filter it, and after filtering (if the data were combined), they would have to have some way to sort out which calls were from which pipeline. I don't think we have that stored in the output MAF file now; the data on things like how many callers agreed is in the notebook, but it gets filtered out before writing. (This is another potential benefit of keeping a separate file: we can output this data for the hostpot calls, which might be useful to end users!)

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (jashapiro) [inline analyses/hotspots-detection/01-create-hotspot-maf.Rmd:44]: A heads up: with the changes in MAF files coming with normalization (https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/990), the set of column headers is changing, so this data file will need to be regenerated as some of the columns are missing. 

Note also that there are unique columns for each caller (see https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/990#issuecomment-826879406) that may be best filtered out in `00-subset-maf.R`

**Comment by jharenza** [comment]
@jashapiro are we good to go with merging this, after the branch is updated?

**Comment by jharenza** [comment]
Based on offline discussion, we created a hotspots only maf to include _all_ hotspots instead of those newly recovered. To the point of subtyping with the hotspot only maf- this will be a case by case decision, as some modules require kinase domain SNVs or duplications, which require the consensus maf.

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd:105]: I think I did not commit the code changes, I'm working on it again now

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd:44]: This is updated as of #1050

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd:115]: caller_count is the number of callers (strelka2,mutect2,vardict,lancet) that called the variant.

Comment (jashapiro) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd:115]: Okay, I think I was confused by the placement of this chunk, since the `caller_counts` dataframe isn't used until much later. 

I was originally going to make a suggestion to not use `toString()` here, because it makes string separated on ", " that is a bit harder to parse to individual entries if that becomes needed. I'd lean toward a bare comma with no space or a semicolon:

```
caller = paste(unique(caller), collapse=';')
```

Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd:104]: ok I'll add that and move it closer to the plotting chunks

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: Now that we have to gather all the hotspots calls in a maf, this step https://github.com/kgaonkar6/OpenPBTA-analysis/blob/ab9c3b2ba7f093c3e1254093fc92ac01d83ed28b/analyses/hotspots-detection/01-create-hotspot-maf.Rmd#L130-L169 takes a lot of RAM so I'm running in m5 instance. I'll  report back after a rerun about the vardict -only calls and updated file.

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: It actually needs to be here since in the next steps the `caller` columns is uniqued by prefix of caller names "strelka2","mutect2","vardict" and "lancet" . I did look into `unite` https://tidyr.tidyverse.org/reference/unite.html to paste the `<caller_name>_caller`  columns using  but that adds NA for columns so that would not be ideal

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: The above RAM issue was because if had multiple rows in *RDS files that were being merged in this script. I have added unique() to each output from filterMaf.R which as solved the issue and now can be run locally without much RAM requirements.


The 79 calls unique to 1 caller was from mafs in v18 release. In v19 , 148 calls are unique to 1 caller and this is the distribution of calls that are unique to just 1 caller:
caller<chr> | n<int>
-- | --
lancet | 38
mutect2 | 24
strelka2 | 9
vardict | 77

[unique_to_1_caller_hotspots.txt](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/6467210/unique_to_1_caller_hotspots.txt)

Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: This is the distribution of calls unique to 1 caller : 
Rplot02

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: This is updated in the latest commit

File: analyses/hotspots-detection/01-create-hotspot-maf.Rmd
Comment (jharenza) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: @kgaonkar6 can you check whether the HLA-A calls are from one caller or multiple?

>I think I asked this before, but if implement a consistent rule that we require >= 2 callers with no real loss, that seems simpler & more consistent with what I have seen elsewhere.

@jashapiro given these comparisons, I wonder if we should implement a >=2 callers if vardict is is included or a >=1 caller if it is from one of the other 3?

Comment (jharenza) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: Another option could be adding all of these and having a column for caller, such that users can filter out if just from VarDict, for example.

Comment (jharenza) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: I dug into HLA-A, because it seemed spurious to me. It looks like the mutations were called by both Mutect2 and Lancet, and for each sample, there are two adjacent mutations. Turns out, this is an [MNV](https://www.ncbi.nlm.nih.gov/snp/?term=rs796080364) in dbSNP. It is benign in PolyPhen, so I am wondering if this should even be a real hotspot.

Comment (kgaonkar6) [reply analyses/hotspots-detection/01-create-hotspot-maf.Rmd]: I've added all the calls that are unique to 1 caller in the notebook for review as suggested by @jharenza

**Review by jharenza — APPROVED** [review]
This looks good to be merged. Note: the one caller only table is now in the notebook for easy viewing.

**Comment by jharenza** [comment]
I am going to merge this so we can get to #959

**Pull request merged**
